File: packages/core/src/data/csv.ts
/**
 * A tiny dependency-free CSV parser + typed-array column access — the common
 * "load a CSV and plot a column" path. Handles quoted fields, escaped quotes
 * (`""`), and `\n` / `\r\n` line endings.
 */

export interface CSVOptions {
  /** Field delimiter. Default `","`. */
  delimiter?: string;
  /** Treat the first row as headers. Default `true`. */
  header?: boolean;
  /** Drop blank lines. Default `true`. */
  skipEmpty?: boolean;
}

/** A parsed CSV: header names + string rows, with typed column accessors. */
export interface Table {
  headers: string[];
  rows: string[][];
  /** Number of data rows. */
  readonly length: number;
  /** A column's raw string values (by name or index). */
  column(name: string | number): string[];
  /** A column parsed to a `Float64Array` (non-numeric cells become `NaN`). */
  numeric(name: string | number): Float64Array;
}

function tokenize(text: string, delim: string): string[][] {
  const rows: string[][] = [];
  let row: string[] = [];
  let field = "";
  let inQuotes = false;
  for (let i = 0; i < text.length; i++) {
    const c = text[i]!;
    if (inQuotes) {
      if (c === '"') {
        if (text[i + 1] === '"') { field += '"'; i++; } else inQuotes = false;
      } else field += c;
      continue;
    }
    if (c === '"') { inQuotes = true; }
    else if (c === delim) { row.push(field); field = ""; }
    else if (c === "\n" || c === "\r") {
      if (c === "\r" && text[i + 1] === "\n") i++;
      row.push(field); rows.push(row); row = []; field = "";
    } else field += c;
  }
  if (field.length > 0 || row.length > 0) { row.push(field); rows.push(row); }
  return rows;
}

/** Parse CSV text into a {@link Table}. */
export function parseCSV(text: string, opts: CSVOptions = {}): Table {
  const delim = opts.delimiter ?? ",";
  const useHeader = opts.header !== false;
  let all = tokenize(text, delim);
  if (opts.skipEmpty !== false) all = all.filter((r) => !(r.length === 1 && r[0]!.trim() === ""));

  let headers: string[];
  let dataRows: string[][];
  if (useHeader && all.length) { headers = all[0]!.map((h) => h.trim()); dataRows = all.slice(1); }
  else { const n = all[0]?.length ?? 0; headers = Array.from({ length: n }, (_, i) => `col${i}`); dataRows = all; }

  const idx = (name: string | number): number => (typeof name === "number" ? name : headers.indexOf(name));
  return {
    headers,
    rows: dataRows,
    get length() { return dataRows.length; },
    column(name) {
      const i = idx(name);
      return dataRows.map((r) => r[i] ?? "");
    },
    numeric(name) {
      const i = idx(name);
      const out = new Float64Array(dataRows.length);
      for (let k = 0; k < dataRows.length; k++) out[k] = Number.parseFloat(dataRows[k]![i] ?? "");
      return out;
    },
  };
}

File: native/src/data/csv.cpp
#include "data/csv.hpp"

#include <algorithm>
#include <cstdlib>
#include <limits>

namespace photon::data {

namespace {

const std::string kEmpty;

std::vector<std::vector<std::string>> tokenize(const char* text, size_t length, char delim) {
  std::vector<std::vector<std::string>> rows;
  std::vector<std::string> row;
  std::string field;
  bool in_quotes = false;
  for (size_t i = 0; i < length; ++i) {
    const char c = text[i];
    if (in_quotes) {
      if (c == '"') {
        // A doubled quote inside a quoted field is one literal quote.
        if (i + 1 < length && text[i + 1] == '"') {
          field.push_back('"');
          ++i;
        } else {
          in_quotes = false;
        }
      } else {
        field.push_back(c);
      }
      continue;
    }
    if (c == '"') {
      in_quotes = true;
    } else if (c == delim) {
      row.push_back(field);
      field.clear();
    } else if (c == '\n' || c == '\r') {
      if (c == '\r' && i + 1 < length && text[i + 1] == '\n') ++i;
      row.push_back(field);
      rows.push_back(row);
      row.clear();
      field.clear();
    } else {
      field.push_back(c);
    }
  }
  if (!field.empty() || !row.empty()) {
    row.push_back(field);
    rows.push_back(row);
  }
  return rows;
}

std::string trim(const std::string& s) {
  size_t begin = 0;
  size_t end = s.size();
  const auto space = [](char c) {
    return c == ' ' || c == '\t' || c == '\n' || c == '\r' || c == '\f' || c == '\v';
  };
  while (begin < end && space(s[begin])) ++begin;
  while (end > begin && space(s[end - 1])) --end;
  return s.substr(begin, end - begin);
}

}  // namespace

int Table::index_of(const std::string& name) const {
  for (size_t i = 0; i < headers_.size(); ++i) {
    if (headers_[i] == name) return static_cast<int>(i);
  }
  return -1;
}

const std::string& Table::cell(size_t row, size_t column) const {
  if (row >= rows_.size() || column >= rows_[row].size()) return kEmpty;
  return rows_[row][column];
}

std::vector<double> Table::numeric(size_t column) const {
  std::vector<double> out(rows_.size(), std::numeric_limits<double>::quiet_NaN());
  for (size_t r = 0; r < rows_.size(); ++r) {
    const std::string& text = cell(r, column);
    if (text.empty()) continue;
    // strtod, like parseFloat, reads a leading number and ignores the tail; a
    // cell with no number at all leaves `end` at the start and stays NaN.
    char* end = nullptr;
    const double v = std::strtod(text.c_str(), &end);
    if (end != text.c_str()) out[r] = v;
  }
  return out;
}

Table parse_csv(const char* text, size_t length, const CsvOptions& opts) {
  if (!text) return Table();
  std::vector<std::vector<std::string>> all =
      tokenize(text, length, opts.delimiter == '\0' ? ',' : opts.delimiter);
  if (opts.skip_empty) {
    all.erase(std::remove_if(all.begin(), all.end(),
                             [](const std::vector<std::string>& r) {
                               return r.size() == 1 && trim(r[0]).empty();
                             }),
              all.end());
  }

  std::vector<std::string> headers;
  std::vector<std::vector<std::string>> rows;
  if (opts.header && !all.empty()) {
    headers.reserve(all[0].size());
    for (const std::string& h : all[0]) headers.push_back(trim(h));
    rows.assign(all.begin() + 1, all.end());
  } else {
    const size_t n = all.empty() ? 0 : all[0].size();
    headers.reserve(n);
    for (size_t i = 0; i < n; ++i) headers.push_back("col" + std::to_string(i));
    rows = std::move(all);
  }
  return Table(std::move(headers), std::move(rows));
}

}  // namespace photon::data
